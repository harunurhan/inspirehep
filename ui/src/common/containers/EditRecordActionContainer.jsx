import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Set } from 'immutable';
import { doSetsHaveCommonItem } from '../utils';
import ListItemAction from '../../common/components/ListItemAction';

const ONLY_SUPER_USERS_AND_CATALOGERS = Set(['superuser', 'cataloger']);

class EditRecordActionContainer extends Component {
  render() {
    const { userRoles, recordId } = this.props;
    const isAuthorized = doSetsHaveCommonItem(
      userRoles,
      ONLY_SUPER_USERS_AND_CATALOGERS
    );
    const href = `/workflows/edit_article/${recordId}`;
    if (isAuthorized) {
      return (
        <ListItemAction
          iconType="edit"
          text="edit"
          link={{
            href,
            target: '_blank',
          }}
        />
      );
    }
    return null;
  }
}

EditRecordActionContainer.propTypes = {
  userRoles: PropTypes.instanceOf(Set).isRequired,
  recordId: PropTypes.number.isRequired,
};

const stateToProps = state => ({
  userRoles: Set(state.user.getIn(['data', 'roles'])),
});

export default connect(stateToProps)(EditRecordActionContainer);
