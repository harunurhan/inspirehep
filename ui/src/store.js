import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import createHistory from 'history/createBrowserHistory';
import { routerMiddleware } from 'react-router-redux';
/* eslint-disable import/no-extraneous-dependencies */
import { createLogger } from 'redux-logger';
/* eslint-disable import/no-extraneous-dependencies */
import { composeWithDevTools } from 'redux-devtools-extension';

import reducers from './reducers';
import http from './common/http';
import queryParamsParserMiddleware from './middlewares/queryParamsParser';
import keepPreviousUrlMiddleware from './middlewares/keepPreviousUrl';
import {
  reHydrateRootStateFromStorage,
  createPersistToStorageMiddleware,
} from './middlewares/statePersister';
import searchDispatcherMiddleware from './middlewares/searchDispatcher';
import redirectToErrorPageMiddleware from './middlewares/redirectToErrorPage';
import actionTrackerMiddleware from './middlewares/actionTracker';

export const thunkMiddleware = thunk.withExtraArgument(http);

export const history = createHistory();
const reduxRouterMiddleware = routerMiddleware(history);

const reducersToPersist = ['ui', 'user'];

const PROD_MIDDLEWARES = [
  reduxRouterMiddleware,
  queryParamsParserMiddleware,
  keepPreviousUrlMiddleware,
  createPersistToStorageMiddleware(reducersToPersist),
  searchDispatcherMiddleware,
  redirectToErrorPageMiddleware,
  actionTrackerMiddleware,
  thunkMiddleware,
];

const DEV_MIDDLEWARES = [...PROD_MIDDLEWARES, createLogger()];

const withMiddlewares = () => {
  if (process.env.NODE_ENV === 'production') {
    return applyMiddleware(...PROD_MIDDLEWARES);
  }
  return applyMiddleware(...DEV_MIDDLEWARES);
};

export const store = createStore(
  reducers,
  reHydrateRootStateFromStorage(reducersToPersist),
  composeWithDevTools(withMiddlewares())
);
