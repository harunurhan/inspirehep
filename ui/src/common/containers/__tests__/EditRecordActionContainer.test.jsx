import React from 'react';
import { fromJS } from 'immutable';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { getStoreWithState } from '../../../fixtures/store';
import EditRecordActionContainer from '../EditRecordActionContainer';

describe('EditRecordActionContainer', () => {
  it('renders edit button if user is superuser', () => {
    const store = getStoreWithState({
      user: fromJS({
        data: {
          roles: ['superuser'],
        },
      }),
    });
    const wrapper = mount(
      <Provider store={store}>
        <EditRecordActionContainer recordId={123456} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders edit button if user is cataloger', () => {
    const store = getStoreWithState({
      user: fromJS({
        data: {
          roles: ['cataloger'],
        },
      }),
    });
    const wrapper = mount(
      <Provider store={store}>
        <EditRecordActionContainer recordId={123456} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('does not render edit button if user is not cataloger or superuser', () => {
    const store = getStoreWithState({
      user: fromJS({
        data: {
          roles: ['unauthorized'],
        },
      }),
    });
    const wrapper = mount(
      <Provider store={store}>
        <EditRecordActionContainer recordId={123456} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
