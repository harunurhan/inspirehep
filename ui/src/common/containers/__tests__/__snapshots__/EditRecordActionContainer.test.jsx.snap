// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditRecordActionContainer does not render edit button if user is not cataloger or superuser 1`] = `""`;

exports[`EditRecordActionContainer renders edit button if user is cataloger 1`] = `
<a
  className="ant-btn __ListItemAction__"
  href="/workflows/edit_article/123456"
  onClick={[Function]}
  target="_blank"
>
  <span
    className="__IconText__"
  >
    <i
      className="anticon anticon-edit icon"
    />
    edit
  </span>
</a>
`;

exports[`EditRecordActionContainer renders edit button if user is superuser 1`] = `
<a
  className="ant-btn __ListItemAction__"
  href="/workflows/edit_article/123456"
  onClick={[Function]}
  target="_blank"
>
  <span
    className="__IconText__"
  >
    <i
      className="anticon anticon-edit icon"
    />
    edit
  </span>
</a>
`;
